### PR TITLE
Initial implementation of learning-based PEC

### LINKED ISSUE #821: New technique: Learning-based quantum error mitigation
### Problem

Mitiq needs more error mitigation techniques.

The [Learning-based quantum error mitigation](https://arxiv.org/abs/2005.07601) could be viewed as an enhancement to `mitiq.cdr`: Fit data with ansatz other than ax + b.

### Rough requirements

- TBD

## Draft RFC
https://docs.google.com/document/d/1VrZRV59-2ApemebfbL556Rj7I2AMhAS7bJgGuTg3K1w/edit?usp=sharing

### LINKED ISSUE #1270: Create functions to learn biased noise
Pre-Request Checklist
---------------------

- [x] I checked to make sure that this feature has not already been requested.

Issue Description
-----------------

Add file to codebase to create functions to learn biased noise

Proposed Solution
-----------------

Let us know what you'd like to see happen, and be sure to mention
any alternatives that you've considered.

Additional References
---------------------

If applicable, provide some references that will help us better understand the request.

### PR DESCRIPTION
**Pull request by Misty-W**
Create file for learning-based PEC

Description
-----------

Fixes #1270. 

Checklist
---------

Check off the following once complete (or if not applicable) after opening the PR. The PR will be reviewed once this checklist is complete and all tests are passing.

- [x] I added unit tests for new code.
- [x] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [x] I used [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) docstrings for functions.
- [x] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.
- [X] Added myself / the copyright holder to the AUTHORS file

If some items remain, you can mark this a [draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

License
-------

- [X] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

Tips
----

- If the validation check fails:

    1. Run `make check-types` (from the root directory of the repository) and fix any [mypy](https://mypy.readthedocs.io/en/stable/) errors.

    2. Run `make check-style` and fix any [flake8](http://flake8.pycqa.org) errors.

    3. Run `make format` to format your code with the [black](https://black.readthedocs.io/en/stable/index.html) autoformatter.

  For more information, check the [Mitiq style guidelines](https://mitiq.readthedocs.io/en/stable/contributing.html#style-guidelines).

- Write "Fixes #XYZ" in the description if this PR fixes Issue #XYZ.

### CODE REVIEW

**Comment by Misty-W** [comment]
Very much "under construction" at the moment 🙂

File: mitiq/pec/representations/learn_noise_parameters.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+from scipy.optimize import minimize
+from mitiq import QPROGRAM, Executor, Observable
+from mitiq.cdr import generate_training_circuits
+from mitiq.pec import execute_with_pec
+from mitiq.pec.representations.biased_noise import (
+    represent_operation_with_local_biased_noise,
+)
+
+
+def learn_noise_parameters(
```
Comment (andreamari) [inline mitiq/pec/representations/learn_noise_parameters.py:29]:
```suggestion
def learn_biased_noise_parameters(
```
We may want to learn other noise models in the future

File: mitiq/pec/representations/learn_noise_parameters.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
+# Copyright (C) 2022 Unitary Fund
```
Comment (andreamari) [inline mitiq/pec/representations/learn_noise_parameters.py:1]: The name of this file `learn_noise_parameters.py` looks more like a function name than a file name.
I would suggest a simpler and shorter name for the file. 
E.g. `learning.py`

File: mitiq/pec/representations/learn_noise_parameters.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        method="BFGS",
+    )
+    x_result = result.x
+    epsilon = x_result[0]
+    eta = x_result[1]
+
+    return [epsilon, eta]
+
+
+def loss_function(
```
Comment (andreamari) [inline mitiq/pec/representations/learn_noise_parameters.py:85]:
```suggestion
def biased_noise_loss_function(
```
We may add other noise models in the future.

File: mitiq/pec/representations/learn_noise_parameters.py (hunk 2 above)
Comment (Misty-W) [reply mitiq/pec/representations/learn_noise_parameters.py]: good point- changed the filename to `learning.py`

**Comment by nathanshammah** [comment]
For the link check tests failing, I suggest removing the `doi` field from the three quantum journal entries and see what happens.

File: mitiq/pec/representations/tests/test_learning.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    ]
+
+    def ideal_executor(circ: Circuit) -> np.ndarray:
+        return compute_density_matrix(circ, noise_level=(0.0,))
+
+    def noisy_executor(circ: Circuit) -> np.ndarray:
+        qreg = LineQubit.range(gate.num_qubits())
+        return compute_density_matrix(
+            circ, noise_model=ops.MixedUnitaryChannel(mix).on_each(*qreg)
+        )
```
Comment (andreamari) [inline mitiq/pec/representations/tests/test_learning.py:71]:
```suggestion
    def noisy_executor(circ: Circuit) -> np.ndarray:
        noisy_circ = circ.with_noise(ops.MixedUnitaryChannel(mix))
        return ideal_executor(noisy_circ)
```
This change help too. Looks safer with respect to qubits

File: mitiq/pec/representations/learning.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    epsilon = x_result[0]
+    eta = x_result[1]
+
+    return [epsilon, eta]
+
+
+def biased_noise_loss_function(
+    x0: np.ndarray,
+    operation: QPROGRAM,
+    circuit: QPROGRAM,
```
Comment (andreamari) [inline mitiq/pec/representations/learning.py:101]: This is more a conceptual question. Shouldn't `biased_noise_loss_function` take a list of `training_circuits`  as argument, such that we can get the associated list of `mitigated_values` to be compared with `ideal_values`?

At the moment we compare the `mitigated_value` of `circuit` which is not necessarily close to the ideal results of  `training_circuits`

Comment (Misty-W) [reply mitiq/pec/representations/learning.py:100]: The list of `training_circuits` is used to calculate the list of `ideal_values` before the loop (so outside of the loss function). Then the loss function compares the list of `mitigated_values` with the list of `ideal_values`.

The confusion might have come from the `num_training_circuits`  argument of the loss function, which actually isn't needed. Now I'm using `len(ideal_values)` in its place.

**Comment by Misty-W** [comment]
> For the link check tests failing, I suggest removing the `doi` field from the three quantum journal entries and see what happens and/or substitute the doi url in the `url` field with the actual url, like "https://quantum-journal.org/papers/q-2021-11-26-592/".

@nathanshammah, thanks for the suggestion! I propose to track the link issue in https://github.com/unitaryfund/mitiq/issues/1308.

**Comment by Misty-W** [comment]
Split up into smaller pieces: loss function #1340 and learning function (PR to be created once #1340 closed).

**Pull request closed without merging**